### PR TITLE
[ROCm][AITER] Enable fp8 kv cache on rocm aiter backend.

### PR DESCRIPTION
**Pull request by fsx950223**
Rocm aiter backend could support fp8 kv cache with latest aiter
CMD:
HIP_VISIBLE_DEVICES=3,4 VLLM_ROCM_USE_AITER=1 VLLM_USE_V1=1 vllm serve /models/models--amd--Meta-Llama-3.1-8B-Instruct-FP8-KV/snapshots/fa42f9a9105c545755fea25cf69f49ac8c8b40e1/ --tensor-parallel-size 2 --gpu-memory-utilization 0.9 --trust-remote-code --disable-log-requests --block-size 16 --max-model-len 32768 --dtype float16 --quantization fp8 --no-enable-prefix-caching --max-num-batched-tokens=8192 --kv-cache-dtype fp8 --compilation-config '{"full_cuda_graph":true}'

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.796|±  |0.0255|
|     |       |strict-match    |     5|exact_match|↑  |0.744|±  |0.0277|
---------

## Essential Elements of an Effective PR Description Checklist
- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

### CODE REVIEW

File: vllm/v1/attention/backends/rocm_aiter_fa.py [hunk 1]
```diff
@@ -559,28 +591,14 @@ def forward(
                     alibi_slopes=self.alibi_slopes,
                     window_size=self.sliding_window,
                     block_table=block_table,
-                    cu_seqlens_k=(cu_seq_lens if not use_local_attn else
-                                  local_metadata.local_cu_seq_lens),
+                    cu_seqlens_k=cu_seq_lens,
```
Comment (tjtanaa) [inline vllm/v1/attention/backends/rocm_aiter_fa.py:503]: @fsx950223 can you double check if there is accidental removal of a bug fix for local attention cases here?

it should be using the `local_metadata.local_cu_seq_lens` when `local_attention` is true.

There is alignment issue when processing an input sequence larger than 8192.

The bug fix PR for local attention was introduced here https://github.com/vllm-project/vllm/pull/19904

Comment (fsx950223) [reply vllm/v1/attention/backends/rocm_aiter_fa.py:503]: Done

File: vllm/attention/ops/chunked_prefill_paged_decode.py [hunk 2]
```diff
@@ -305,7 +304,7 @@ def chunked_prefill_paged_decode(
         )
         max_logits = torch.empty_like(exp_sums)
 
-        ops.paged_attention_rocm(
+        torch.ops.aiter.paged_attention_rocm(
```
Comment (tjtanaa) [inline vllm/attention/ops/chunked_prefill_paged_decode.py]: @fsx950223 I think we should avoid replacing this directly.

vLLM's repo `ops.paged_attention_rocm` is more generic and covers more GPU arch e.g. Radeon GPUs.
With this replacement, it will affect users who are using Radeon GPUs. They could not build AITER repo on Radeon. 

CC. @gshtras  @hongxiayang @hyoon1

File: vllm/platforms/rocm.py [hunk 3]
```diff
@@ -138,9 +138,9 @@ def use_rocm_custom_paged_attention(
         return ((not envs.VLLM_USE_V1 or sliding_window == 0
                  or sliding_window == (-1, -1))
                 and (qtype == torch.half or qtype == torch.bfloat16)
-                and (head_size == 64 or head_size == 128)
+                and (head_size in [64, 128, 256])
```
Comment (tjtanaa) [inline vllm/platforms/rocm.py]: @fsx950223 

Based on this comment https://github.com/vllm-project/vllm/pull/20295/files#r2179115840 , as we need to retain the use of `ops.paged_attention_rocm` , we should create another if else branch just to handle the case for the AITER Flash Attention.

File: vllm/attention/ops/chunked_prefill_paged_decode.py (hunk 2 above)
Comment (fsx950223) [reply vllm/attention/ops/chunked_prefill_paged_decode.py]: There should be a switch, for Radeon GPUs, use original one, will change later.

File: vllm/attention/backends/rocm_flash_attn.py [hunk 4]
```diff
@@ -913,8 +913,7 @@ def forward(
                 )
                 max_logits = torch.empty_like(exp_sums)
 
-                query_start_loc = None
-                ops.paged_attention_rocm(
+                torch.ops.aiter.paged_attention_rocm(
```
Comment (tjtanaa) [inline vllm/attention/backends/rocm_flash_attn.py]: @fsx950223 

Same situation as the comment https://github.com/vllm-project/vllm/pull/20295/files#r2179115840 .

File: vllm/attention/ops/chunked_prefill_paged_decode.py (hunk 2 above)
Comment (tjtanaa) [reply vllm/attention/ops/chunked_prefill_paged_decode.py]: Great. Hope to make the landing of this PR smooth and we can benefit from the AITER speed boost. 🙌 🚀

File: vllm/platforms/rocm.py (hunk 3 above)
Comment (fsx950223) [reply vllm/platforms/rocm.py]: The condition only for gfx9xx

File: vllm/attention/backends/rocm_flash_attn.py (hunk 4 above)
Comment (fsx950223) [reply vllm/attention/backends/rocm_flash_attn.py]: Only gfx9xx gpus support rocm flash attn.

Comment (fsx950223) [reply vllm/attention/backends/rocm_flash_attn.py]: We have to change this for v0 workloads.

File: vllm/platforms/rocm.py (hunk 3 above)
Comment (tjtanaa) [reply vllm/platforms/rocm.py]: Is ROCm/aiter compatible with Mi200 series GPU? (e.g. gfx90a) ? On upstream gfx9xx also include Mi200 series GPU support.

File: vllm/attention/backends/rocm_flash_attn.py (hunk 4 above)
Comment (tjtanaa) [reply vllm/attention/backends/rocm_flash_attn.py]: Is ROCm/aiter compatible with Mi200 series GPU? (e.g. gfx90a) ?

File: vllm/platforms/rocm.py (hunk 3 above)
Comment (tjtanaa) [reply vllm/platforms/rocm.py]: Moreover `chunked_prefill_decode attention` relies on this condition as well (it uses `ops.paged_attention_rocm`). `ops.paged_attention_rocm` depends on a different set of conditions than that of `torch.ops.aiter.paged_attention_rocm`.

https://github.com/vllm-project/vllm/blob/d853520b3e47136c3e434cfcfea0473d352f0971/vllm/attention/ops/chunked_prefill_paged_decode.py#L284

So we still need to create a separate condition for the AITER FA

Comment (fsx950223) [reply vllm/platforms/rocm.py]: > Is ROCm/aiter compatible with Mi200 series GPU? (e.g. gfx90a) ? On upstream gfx9xx also include Mi200 series GPU support.

Yes

Comment (fsx950223) [reply vllm/platforms/rocm.py]: > Moreover `chunked_prefill_decode attention` relies on this condition as well (it uses `ops.paged_attention_rocm`). It depends on a different set of conditions than that of `torch.ops.aiter.paged_attention_rocm`.
> 
> https://github.com/vllm-project/vllm/blob/d853520b3e47136c3e434cfcfea0473d352f0971/vllm/attention/ops/chunked_prefill_paged_decode.py#L284
> 
> So we still need to create a separate condition for the AITER FA

Triton backend has its own condition

File: vllm/attention/backends/rocm_flash_attn.py (hunk 4 above)
Comment (fsx950223) [reply vllm/attention/backends/rocm_flash_attn.py]: Yes

File: vllm/platforms/rocm.py (hunk 3 above)
Comment (fsx950223) [reply vllm/platforms/rocm.py]: Make a new limit in the file

Comment (tjtanaa) [reply vllm/platforms/rocm.py]: The limit also needs to include the `gqa_ratio` limits?

latest changes in the code:
```diff
- (gqa_ratio >= 1 and gqa_ratio <= 16)
+ (gqa_ratio >= 1 and gqa_ratio <= 32)
```

Comment (fsx950223) [reply vllm/platforms/rocm.py]: Done

File: vllm/attention/ops/chunked_prefill_paged_decode.py (hunk 2 above)
Comment (gshtras) [reply vllm/attention/ops/chunked_prefill_paged_decode.py]: Not just Radeon. Aiter is not a hard requirement for running vLLM, so any API should be opt-in, not opt-out

**Comment by fsx950223** [comment]
Only keep v1 changes in the PR.

**Comment by hongxiayang** [comment]
Thanks @fsx950223. Can you add test plan in your PR's description section?

**Review by gshtras — APPROVED** [review]

**Comment by tjtanaa** [comment]
Does this feature work with the current aiter commit that vLLM is using? Or do we have to upgrade the aiter? I saw there are a few more new commits on FA and PA feature after the current aiter commit.

**Comment by fsx950223** [comment]
> Does this feature work with the current aiter commit that vLLM is using? Or do we have to upgrade the aiter? I saw there are a few more new commits on FA and PA feature after the current aiter commit.

You could try it.

**Review by SageMoore — APPROVED** [review]
This generally looks reasonable to me. Thanks for the contribution!
